### PR TITLE
Fix misuse of active mask (#98157)

### LINKED ISSUE #98157: [CUDA] `at::native::countRadixUsingMask` misuses `__activemask` intrinsic
### 🐛 Describe the bug

Within `at::native::countRadixUsingMask`, the `__activemask` intrinsic is used to conduct a warp vote (`__ballot_sync(__activemask(), vote)`) between the threads counting the distribution of the radix within the input data.

Since at least CUDA 9, `__activemask` should not be used to determine which threads are along the same execution path as detailed in [this NVIDIA blog post](https://developer.nvidia.com/blog/using-cuda-warp-level-primitives/). Since there is not a guarantee all threads within the same execution path are in the same active thread group, the distribution count results can be off resulting in an wrong assumption of unique results in `at::native::findPattern` which leads to a data race. Can affect `topk`, `kthvalue`, and `median` operators since all three rely on `at::native::radixSelect`.

The issue is hard to reproduce since even though CUDA does not guarantee threads in the same execution path operate within the same active thread group, in practice, this divergence is rare.

Example crash when using `topk`
```python
import torch
x = torch.tensor([0, 1, 2]).cuda()
x.topk(1)
```
```
aten/src/ATen/native/cuda/TensorTopK.cu:144:gatherTopK: assertion: outputSliceSize >= writeIndexStar
```

### Versions

Affects all versions since at least #17544

cc @ngimel

### PR DESCRIPTION
**Pull request by JackWolfard**
Fixes #98157

cc @ngimel

### CODE REVIEW

File: aten/src/ATen/native/cuda/SortingRadixSelect.cuh [hunk 1]
```diff
@@ -213,9 +216,13 @@ __device__ void countRadixUsingMask(
 #if defined(USE_ROCM)
       counts[j] += __popcll(WARP_BALLOT(vote));
 #else
-      counts[j] += __popc(WARP_BALLOT(vote, ACTIVE_MASK()));
+      counts[j] += __popc(WARP_BALLOT(vote, mask));
 #endif
     }
+    i += blockDim.x;
+#if !defined(USE_ROCM)
+    mask = WARP_BALLOT(i < sliceSize, mask);
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/SortingRadixSelect.cuh:224]: is this needlessly complicated? `mask` can be not `0xffffffff` only on the last iteration, where updating it for the next iteration is pointless?

Comment (JackWolfard) [reply aten/src/ATen/native/cuda/SortingRadixSelect.cuh:224]: Assuming only the last iteration will result in a subset of threads exiting the loop early, then `mask` will be used by the subset still in the loop to conduct the ballot. Updating the `mask` on the last iteration is unnecessary, but how is one thread supposed to figure out that it's the last iteration? That's the question I had since I wasn't sure you could rely on this assumption. Is there a scenario where a subset of threads do multiple iterations?

If the assumption holds, then a more efficient solution would be to precompute the number of iterations and for iterations - 1 use a ballot for all threads, while on the last iteration, there would be two ballots. One to determine what threads are still in the iteration, and another to do the normal ballot. The overhead for the extra ballots would be drastically reduced to just one extra ballot instead of one on every iteration.

Comment (ngimel) [reply aten/src/ATen/native/cuda/SortingRadixSelect.cuh:224]: That's right, you could avoid most warp ballots this way but that requires bigger code changes. Without that, could you just leave the code as it was and use `WARP_BALLOT(i < sliceSize)` instead of `ACTIVE_MASK()`? You never need to and predicate with mask I think.

Comment (JackWolfard) [reply aten/src/ATen/native/cuda/SortingRadixSelect.cuh:224]: > [C]ould you just leave the code as it was and use `WARP_BALLOT(i < sliceSize)` instead of `ACTIVE_MASK()`?

No, you can't drop the `mask` since implicitly a full mask will be used. In the case where not all threads can reach a primitive with a full mask, then the program could hang. I got this from [this NVIDIA blogpost](https://developer.nvidia.com/blog/using-cuda-warp-level-primitives/) under Update Legacy Warp-Level Programming point 1.

Comment (ngimel) [reply aten/src/ATen/native/cuda/SortingRadixSelect.cuh:224]: cool, can you leave a comment for the first WARP_BALLOT call then, that it needs to be called outside the loop to make sure all threads participate?

**Review by ngimel — APPROVED** [review]

File: aten/src/ATen/native/cuda/SortingRadixSelect.cuh (hunk 1 above)
Comment (JackWolfard) [reply aten/src/ATen/native/cuda/SortingRadixSelect.cuh:225]: I have added the comment. If you want, I can draft a proposal which just does an extra warp ballot for the last iteration. 

I was thinking it'd look like this somehow

```cuda
index_t iters = ceil_div<index_t>(sliceSize, (index_t) blockDim.x);
index_t i = threadIdx.x,;
for (index_t k = 0; k < iters - 1; i += blockDim.x, k++) {
   ...
   counts[j] += __popc(WARP_BALLOT(vote));
   ...
}
i += blockDim.x;
unsigned mask = WARP_BALLOT(i < sliceSize);
if (i  < sliceSize) {
  ...
  counts[j] += __popc(WARP_BALLOT(vote, mask));
  ...
}
```

Comment (ngimel) [reply aten/src/ATen/native/cuda/SortingRadixSelect.cuh:225]: It's fine, I don't think this kernel is perf-critical

**Review by kit1980 — APPROVED** [review]
